### PR TITLE
Enhancement: `azurerm_private_endpoint ` expose `private_ip_address` as computed attribute

### LINKED ISSUE #5208: Support for private_ip_address attribute on azurerm_private_endpoint resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Thanks for adding Private Link resources, while it is still in preview!
With both Private Link and Private DNS resources, it should now be possible to provision a Private Endpoint for a PaaS resource e.g. SQL Server, and create the DNS to resolve SQL Server privately. 
However, the bit that is missing to accomplish that is relevant attributes in `azurerm_private_endpoint`, so a DNS record can be created. Currently only `id` is exposed, and I propose to add `private_ip_address`.

### New or Affected Resource(s)

* azurerm_private_endpoint

### Potential Terraform Configuration

This provisions a SQL Server, VNet, Private Endpoint and DNS:

```hcl
resource azurerm_resource_group repro {
  name                         = "${var.prefix}-private-endpoint-issue"
  location                     = var.location
}

resource azurerm_virtual_network network {
  name                         = "${var.prefix}-vnet"
  address_space                = ["10.0.0.0/16"]
  location                     = azurerm_resource_group.repro.location
  resource_group_name          = azurerm_resource_group.repro.name
}

resource azurerm_subnet subnet {
  name                         = "${var.prefix}-subnet"
  resource_group_name          = azurerm_resource_group.repro.name
  virtual_network_name         = azurerm_virtual_network.network.name
  address_prefix               = "10.0.1.0/24"
  enforce_private_link_endpoint_network_policies = true
}

resource random_string password {
  length                       = 12
  upper                        = true
  lower                        = true
  number                       = true
  special                      = true
  override_special             = "." 
}

resource azurerm_sql_server sql_server {
  name                         = "${var.prefix}sqlserver"
  resource_group_name          = azurerm_resource_group.repro.name
  location                     = azurerm_resource_group.repro.location
  version                      = "12.0"
  administrator_login          = "dbadmin"
  administrator_login_password = random_string.password.result
}

resource azurerm_private_endpoint endpoint {
  name                         = "${var.prefix}-endpoint"
  resource_group_name          = azurerm_resource_group.repro.name
  location                     = azurerm_resource_group.repro.location
  subnet_id                    = azurerm_subnet.subnet.id

  private_service_connection {
    is_manual_connection       = false
    name                       = "${var.prefix}-endpoint-connection"
    private_connection_resource_id = azurerm_sql_server.sql_server.id
    subresource_names          = ["sqlServer"]
  }
}

resource azurerm_private_dns_zone sql_server_db_dns_zone {
  name                         = "privatelink.database.windows.net"
  resource_group_name          = azurerm_resource_group.repro.name
}

resource azurerm_private_dns_a_record sql_server_dns_record {
  name                         = azurerm_sql_server.sql_server.name
  zone_name                    = azurerm_private_dns_zone.sql_server_db_dns_zone.name
  resource_group_name          = azurerm_resource_group.repro.name
  ttl                          = 300
  # Proposed Attribute for resource azurerm_private_endpoint: private_ip_address
  records                      = [azurerm_private_endpoint.endpoint.private_ip_address]
}

resource azurerm_private_dns_zone_virtual_network_link link {
  name                         = "${azurerm_virtual_network.network.name}-dns"
  resource_group_name          = azurerm_resource_group.repro.name
  private_dns_zone_name        = azurerm_private_dns_zone.sql_server_db_dns_zone.name
  virtual_network_id           = azurerm_virtual_network.network.id
}
```

### LINKED ISSUE #5622: Expose Private IP Address as exported attribute in azurerm_private_endpoint
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
When creating a Private Endpoint, a common use case is to create a Private DNS Zone that overrides the Azure PaaS FQDN with the private IP address from the private endpoint. For example, if I have an Azure Storage Account _mystorageaccount_ and I have a Private Endpoint in my VNET, I would create a Private DNS Zone for _blob.core.windows.net_ and create an DNS A-Record for _mystorageaccount_ with the Private Endpoint private IP address.

When I use _azurerm_private_endpoint_ to create the Private Endpoint, the only [exported attribute](https://www.terraform.io/docs/providers/azurerm/r/private_endpoint.html#attributes-reference) is the Resource ID. To find the private IP address of the private endpoint I need to use a _azurerm_private_endpoint_connection_. Here is where I get into problems. The _azurerm_private_endpoint_connection_ is a data reference, so I need to set a dependency on the _azurerm_private_endpoint_ or there will be a runtime failure as the private endpoint has not been created yet, and the data reference is null. 

By setting the dependency between the two, Terraform will not attempt to reference the _azurerm_private_endpoint_connection_ until it is first used, which in the example I have below will not be until after the _azurerm_private_endpoint_ is created. Because Terraform cannot resolve the _azurerm_private_endpoint_connection_ until the execution phase, it means everytime I run this Terraform configuration, the plan phase says it needs to make a change. If I have the private IP address of the private endpoint in a DNS A-Record, it will update that and override the value with the same IP address each time I run _terraform apply_.

Providing the private IP address as an exported attribute in the _azurerm_private_endpoint_ would remove this problem. 

The following example shows how I can retrieve the private IP address of a private-link enabled AKS cluster.
```
# Create a private endpoint to the private AKS cluster
resource "azurerm_private_endpoint" "ple" {
  name                = "kube-apiserver"
  resource_group_name = "test-rg"
  subnet_id           = data.azurerm_subnet.ple_subnet.id

  private_service_connection {
    name                           = "kube-apiserver"
    private_connection_resource_id = azurerm_kubernetes_cluster.aks.id
    subresource_names              = [ "management" ]
    is_manual_connection           = false
  }
}

# Find the IP Address associated with the private endpoint created above
data "azurerm_private_endpoint_connection" "ple_connection" {
  name                = "kube-apiserver"
  resource_group_name = "test-rg"
  depends_on          = [ azurerm_private_endpoint.ple ]
}

# Create a local variable to the private IP address
locals {
  kube_ple_ip = data.azurerm_private_endpoint_connection.ple_connection.private_service_connection.0.private_ip_address
}

# Create a DNS A-Record using the private IP address
resource "azurerm_private_dns_a_record" "aks_dns_record" {
  name                = "aks-fqdn"
  zone_name           = azurerm_private_dns_zone.build_server_dns_zone.name
  resource_group_name = "test-rg"
  ttl                 = 3600
  records             = [ local.kube_ple_ip ]
}
```

### New or Affected Resource(s)

* azurerm_1.43.0

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

[Private Endpoint exported attribute](https://www.terraform.io/docs/providers/azurerm/r/private_endpoint.html#attributes-reference)

### PR DESCRIPTION
**Pull request by njuCZ**
fixes #5622 
fixes #5208 

add `private_ip_address` just like data source `azurerm_private_endpoint_connection`.

Because `azurerm_private_link_endpoint` use the same flatten function, So I set this field outside this flatten function

### CODE REVIEW

File: azurerm/internal/services/network/resource_arm_private_endpoint.go [hunk 1]
```diff
@@ -193,7 +198,20 @@ func resourceArmPrivateEndpointRead(d *schema.ResourceData, meta interface{}) er
...
+			if nic.ID != nil && *nic.ID != "" {
+				privateIpAddress = getPrivateIpAddress(ctx, nicsClient, *nic.ID)
+			}
+		}
+
 		flattenedConnection := flattenArmPrivateLinkEndpointServiceConnection(props.PrivateLinkServiceConnections, props.ManualPrivateLinkServiceConnections)
+		for _, item := range flattenedConnection {
+			v := item.(map[string]interface{})
+			v["private_ip_address"] = privateIpAddress
+		}
```
Comment (WodansSon) [inline azurerm/internal/services/network/resource_arm_private_endpoint.go]: I don't think you need to loop over the returned slice you just need to append the value to it if the `privateIpAddress` is not empty, something like the below... WDYT?
```
if privateIpAddress != "" {
	privateIp := make(map[string]interface{})
	privateIp["private_ip_address"] = privateIpAddress 
	flattenedConnection = append(flattenedConnection, privateIp)
}
```

Comment (njuCZ) [reply azurerm/internal/services/network/resource_arm_private_endpoint.go]: @WodansSon  the `flattenedConnection` is a list of map, append it with another map will cause `flattenedConnection` has two maps, not a merged map.

Comment (WodansSon) [reply azurerm/internal/services/network/resource_arm_private_endpoint.go]: @njuCZ I got the iteration mixed up in my head... you're right... this is fine.

**Comment by WodansSon** [comment]
image

**Pull request merged**